### PR TITLE
Add API for using an existing producer

### PR DESCRIPTION
**Pull request by iravid**
This continues @13h3r's #197 and incorporates @patriknw and @13h3r's comments.

I rebased this on master and reverted the commit that removes the separate dispatcher again.

### CODE REVIEW

**Comment by iravid** [comment]
Another comment: I continued this branch in the spirit of the original PR, but I find that passing the provider function and a boolean that describes the nature of the provider function to `ProducerStage` somewhat obscure.

Perhaps it would be wiser to create a sealed trait that describes what is being passed:
```
sealed trait ProducerProvider
case class SharedProducer(producer: KafkaProducer) extends ProducerProvider
case class ProducerFactory(factory: () => KafkaProducer) extends ProducerProvider
```

Let me know if you think this is a better approach and I'll adapt the PR.

File: core/src/main/scala/akka/kafka/internal/ProducerStage.scala [hunk 1]
```diff
@@ -19,7 +21,8 @@ import java.util.concurrent.atomic.AtomicInteger
  * INTERNAL API
  */
 private[kafka] class ProducerStage[K, V, P](
-  settings: ProducerSettings[K, V], producerProvider: () => KafkaProducer[K, V]
+  closeTimeout: FiniteDuration,
+  producerProvider: () => KafkaProducer[K, V], closeProducerOnStop: Boolean
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ProducerStage.scala]: minor subjective nitpick: I'd prefer to have `closeProducerOnStop` after `closeTimeout`

Comment (iravid) [inline core/src/main/scala/akka/kafka/internal/ProducerStage.scala]: Agreed - both are configuration parameters. Coming right up...

**Comment by iravid** [comment]
Let me know if you'd like me to squash the commits once review is done

**Pull request merged**